### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by albernsrya**
#Built-in function `len` used as condition

Using the len function to check if a sequence is empty is not idiomatic and can be less performant than checking the truthiness of the object.
len doesn't know the context in which it is called, so if computing the length means traversing the entire sequence, it must; it doesn't know that the result is just being compared to 0. Computing the boolean value can stop after it sees the first element, regardless of how long the sequence actually is.
Unnecessary use of `comprehension`
It is unnecessary to use a comprehension just to loop over the iterable and create a list/set/dict out of it. Python has a specialized set of tools for this task: the list/set/dict constructors, which are faster and more readable.

#Use of `len(seq) - 1` to get last element of an iterable

There’s no need to calculate length of an iterable in order to fetch the last element of the iterable. You can provide a negative index -1 to it directly in orger to get the last element. In this way, you don't have to iterate over the sequence using len to get the last index when your purpose is only to get the last element.

#`hasattr` used to check if the object is callable

It is unreliable to use hasattr(object, '__call__') to check if an object is callabe as it can give wrong results if object implements custom __getattr__ or its __call__ is itself not callabe. It is recommended to use the callable built-in instead.

#Use of an unneeded `not` detected

Please make sure that you can tick the following items at latest before
merging the pull request:

- [ X] added description what has been changed and why
- [X ] if appropriate, include changes into the changelog
- [ X] if appropriate, adapt/extend the documentation
- [X ] as external contributor, you need to sign the Contributor License Agreement (CLA), which can be found in the root directory of the repository

For reviewers:
- [X ] after merging the pull request, check if everything looks like expected (especially the changelog)

### CODE REVIEW

**Review by franzlst — APPROVED** [review]
Thanks a lot @albernsrya. I had a look at your changes and they are all very reasonable. Some of them made me think "how could we ever..." :-D

Any objections @sebastian-brunner?

**Comment by sebastian-brunner** [comment]
No objections, thanks for the mr @albernsrya and your improvements.

File: source/rafcon/gui/controllers/utils/extended_controller.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def __register_actions_of_child_controllers(self):
         assert isinstance(self.__child_controllers, dict)
         for controller in self.__child_controllers.values():
             register_function = getattr(controller, "register_actions", None)
-            if hasattr(register_function, '__call__'):
+             if callable(register_function):
```
Comment (sebastian-brunner) [inline source/rafcon/gui/controllers/utils/extended_controller.py:178]: There is a wrong indent here.

**Comment by sebastian-brunner** [comment]
@franzlst : we probably need to setup a public CI pipeline such that contributors immediately see if they introduced new bugs. I just found above error by pushing albernsrya's branch to our internal CI pipeline.

**Comment by franzlst** [comment]
@albernsrya I fixed theindentation in your `develop` branch, but apparently you need to push it here again or so. You said you signed the CLA, where can we find it? Can you mail it to us?

@sebastian-brunner Indeed, that would be really good. I think this could be achieved using [GitHub actions](https://docs.github.com/en/actions/guides/building-and-testing-python)

**Comment by albernsrya** [comment]
> Thanks a lot @albernsrya. I had a look at your changes and they are all very reasonable. Some of them made me think "how could we ever..." :-D
> 
> Any objections @sebastian-brunner?

thank you for your reply

**Comment by albernsrya** [comment]
> Thanks for the mr @albernsrya and your improvements. Could you fix the remaining issue (please see my comment)?

thanks for your correction, I missed a bit

**Comment by albernsrya** [comment]
> @albernsrya You said you signed the CLA, where can we find it? Can you mail it to us?
> 
> @sebastian-brunner Indeed, that would be really good. I think this could be achieved using [GitHub actions](https://docs.github.com/en/actions/guides/building-and-testing-python)

I already sent my CLA thank you

**Comment by franzlst** [comment]
Thank you again, I received the CLA and merged in your branch.

**Pull request merged**